File: hash_table/133_clone_graph.py
# https://leetcode.com/problems/clone-graph/

from collections import deque

"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    def cloneGraph(self, node: Node) -> Node:

        if not node: return node

        dict = {node.val: Node(node.val, [])}

        queue = deque([node])
        while queue:
            curr = queue.popleft()
            for neighbor in curr.neighbors:
                if neighbor.val not in dict:
                    dict[neighbor.val] = Node(neighbor.val, [])
                    queue.append(neighbor)

                dict[curr.val].neighbors.append(dict[neighbor.val])
        return dict[node.val]

File: hash_table/133_clone_graph.java
package hash_table;

// https://leetcode.com/problems/clone-graph/solutions/3392175/easy-solutions-in-java-python-and-c-look-at-once-with-exaplanation/

import java.util.*;

class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}

class Solution {
    public Node cloneGraph(Node node) {

        // BFS

        // Special Case
        if (node == null) return null;
        
        Node newNode = new Node(node.val);
        HashMap<Integer, Node> map = new HashMap<>();
        map.put(newNode.val, newNode);

        LinkedList<Node> queue = new LinkedList<>();
        queue.add(node);

        while (!queue.isEmpty()){
            Node n = queue.pop();
            for (Node neighbor : n.neighbors){
                if (!map.containsKey(neighbor.val)){
                    map.put(neighbor.val, new Node(neighbor.val));
                    queue.add(neighbor);
                }
                map.get(n.val).neighbors.add(map.get(neighbor.val));
            }
        }
        return newNode;
    }

    
}